### PR TITLE
Enhance shape (#60335)

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#62766 Enhance shape (#60335)**
* #62765 [ONNX] Suppport torch.dot and torch.nn.utils.spectral_norm (#62596)
* #62764 [ONNX] Update repeat_interleave for dynamic repeats (#59979)
* #62763 [ONNX] Fix an issue that optimizations might adjust graph inputs unexpectedly. (#61280)
* #62762 [ONNX] Fix controlflow shape inference with contrib op (#60707)
* #62761 [ONNX] add test images to repo (#61553)
* #62760 [ONNX] Fix for batchnorm training op mode (#52758)
* #62759 [ONNX] Remove aten parameter (#61652)
* #62758 [ONNX] Add support for opset14 in PT-ONNX exporter (#59486)
* #62757 [ONNX] Support lstm_cell symbolic (#61476)

Enhanced shape inference by introducing typeReliableMap.

Co-authored-by: jiafatom <jiafa@microsoft.com>

### CODE REVIEW

**Comment by BowenBao** [comment]
trying to reopen ..
